agent/download.go

```diff
@@ -73,38 +74,43 @@ func (d Download) Start(ctx context.Context) error {
 	})
 }
 
-func getTargetPath(path string, destination string) string {
+func targetPath(ctx context.Context, dlPath, destPath string) string {
+	dlPath = filepath.Clean(dlPath)
+
 	// If we're downloading a file with a path of "pkg/foo.txt" to a folder
 	// called "pkg", we should merge the two paths together. So, instead of it
 	// downloading to: destination/pkg/pkg/foo.txt, it will just download to
 	// destination/pkg/foo.txt
-	destinationPaths := strings.Split(destination, string(os.PathSeparator))
-	downloadPaths := strings.Split(path, string(os.PathSeparator))
-
-	for i := 0; i < len(downloadPaths); i += 100 {
-		// If the last part of the destination path matches
-		// this path in the download, then cut it out.
-		lastIndex := len(destinationPaths) - 1
-
-		// Break if we've gone too far.
-		if lastIndex == -1 {
-			break
-		}
+	destPathComponents := strings.Split(destPath, string(os.PathSeparator))
+	dlPathComponents := strings.Split(dlPath, string(os.PathSeparator))
+
+	// If the last component of the destination path matches the first component
+	// of the download path, then trim the last component of the destination.
+	lastIndex := len(destPathComponents) - 1
+	lastDestComponent := destPathComponents[lastIndex]
+	if lastDestComponent == dlPathComponents[0] {
+		destPathComponents = destPathComponents[:lastIndex]
+		destPath = strings.Join(destPathComponents, string(os.PathSeparator))
+	}
 
-		lastPathInDestination := destinationPaths[lastIndex]
-		if lastPathInDestination == downloadPaths[i] {
-			destinationPaths = destinationPaths[:lastIndex]
+	if experiments.IsEnabled(ctx, experiments.DisallowArtifactPathTraversal) {
+		// Trim empty or ".." components from the prefix of dlPath; it should not be
+		// permitted to walk up the directory tree from destPath.
+		for len(dlPathComponents) > 0 {
+			if c := dlPathComponents[0]; c != "" && c != ".." {
+				break
+			}
+			dlPathComponents = dlPathComponents[1:]
```

Comment:
I feel mutating users' intention is a bit questionable. It's likely to cause some support tickets. 🤔 

And given they can add relative path in both download path and destination path, is this fix sufficient?

Reply:
I answered this out of band, but basically, it's less surprising if a user has deliberately written the command:

```shell
buildkite-agent artifact download "*foo" ../../dest
```

than if they wrote the following (and "*foo" happened to match an artifact with path "../../dest"):

```shell
buildkite-agent artifact download "*foo" .
```